codes/python/chapter_tree/avl_tree.py

```diff
@@ -0,0 +1,281 @@
+import sys, os.path as osp
+import typing
+
+sys.path.append(osp.dirname(osp.dirname(osp.abspath(__file__))))
+from include import *
+
+
+class AVLTreeNode:
+    def __init__(
+        self,
+        val=None,
+        height: int = 0,
+        left: typing.Optional["AVLTreeNode"] = None,
+        right: typing.Optional["AVLTreeNode"] = None,
+    ):
+        self.val = val
+        self.height = height
+        self.left = left
+        self.right = right
+
+    def __str__(self):
+        val = self.val
+        left_val = self.left.val if self.left else None
+        right_val = self.right.val if self.right else None
+        return "<AVLTreeNode: {}, leftAVLTreeNode: {}, rightAVLTreeNode: {}>".format(
+            val, left_val, right_val
+        )
+
+
+class AVLTree:
+    def __init__(self, root: typing.Optional[AVLTreeNode] = None):
+        self.root = root
+
+    @staticmethod
+    def height(node: typing.Optional[AVLTreeNode]) -> int:
+        """
+        获取结点高度
+        Args:
+            node:起始结点
+
+        Returns: 高度 or -1
+
+        """
+        # 空结点高度为 -1 ，叶结点高度为 0
+        if node is not None:
+            return node.height
+        return -1
+
+    def __update_height(self, node: AVLTreeNode):
+        """
+        更新结点高度
+        Args:
+            node: 要更新高度的结点
+
+        Returns: None
+
+        """
+        # 结点高度等于最高子树高度 + 1
+        node.height = max([self.height(node.left), self.height(node.right)]) + 1
+
+    def balance_factor(self, node: AVLTreeNode) -> int:
+        """
+        获取结点平衡因子
+        Args:
+            node: 要获取平衡因子的结点
+
+        Returns: 平衡因子
+
+        """
+        # 空结点平衡因子为 0
+        if node is None:
+            return 0
+        # 结点平衡因子 = 左子树高度 - 右子树高度
+        return self.height(node.left) - self.height(node.right)
+
+    def __right_rotate(self, node: AVLTreeNode) -> AVLTreeNode:
+        child = node.left
+        grand_child = child.right
+        # 以 child 为原点，将 node 向右旋转
+        child.right = node
+        node.left = grand_child
+        # 更新结点高度
+        self.__update_height(node)
+        self.__update_height(child)
+        # 返回旋转后子树的根节点
+        return child
+
+    def __left_rotate(self, node: AVLTreeNode) -> AVLTreeNode:
+        child = node.right
+        grand_child = child.left
+        # 以 child 为原点，将 node 向左旋转
+        child.left = node
+        node.right = grand_child
+        # 更新结点高度
+        self.__update_height(node)
+        self.__update_height(child)
+        # 返回旋转后子树的根节点
+        return child
+
+    def rotate(self, node: AVLTreeNode):
```

Comment:
Please keep the consistency. `def rotate(self, node: AVLTreeNode) -> AVLTreeNode:`